conductor-client/src/main/kotlin/com/openlattice/assembler/PostgresRoles.kt

```diff
@@ -19,28 +20,59 @@ class PostgresRoles private constructor() {
     companion object {
 
         @JvmStatic
-        fun getOrCreatePermissionRole(
-                externalRoleNames: IMap<AccessTarget, UUID>,
-                tableId: UUID,
-                columnId: UUID,
-                permission: Permission
-        ): String {
-            return getOrCreatePermissionRole(externalRoleNames, AccessTarget.forPermissionOnTarget(permission, tableId, columnId))
-        }
+        fun getOrCreatePermissionRolesAsync(
+                permissionRoleNames: IMap<AccessTarget, UUID>,
+                targets: Set<AccessTarget>,
+                orgDataSource: HikariDataSource
+        ): CompletionStage<Map<AccessTarget, UUID>> {
+            val allExistingRoleNames = permissionRoleNames.getAll(targets)
+            val targetsToNewIds = targets.filterNot {
+                // filter out roles that already exist
+                allExistingRoleNames.containsKey(it)
+            }.associateWith {
+                val newRole = UUID.randomUUID()
+                newRole
+            }
 
-        @JvmStatic
-        fun getOrCreatePermissionRole( permissionRoleNames: IMap<AccessTarget, UUID>, column: TableColumn, permission: Permission): String {
-            return getOrCreatePermissionRole(permissionRoleNames, AccessTarget.forPermissionOnTarget(permission, column.tableId, column.columnId))
-        }
+            orgDataSource.connection.use { conn ->
+                conn.createStatement().use { stmt ->
+                    targetsToNewIds.values.forEach { roleName ->
+                        stmt.execute( createExternalDatabaseRoleIfNotExistsSql(roleName.toString()))
+                    }
+                }
+            }
 
-        @JvmStatic
-        fun getOrCreatePermissionRole( permissionRoleNames: IMap<AccessTarget, UUID>, permission: Permission, root: UUID, vararg parts: UUID ): String {
-            return getOrCreatePermissionRole(permissionRoleNames, AccessTarget.forPermissionOnTarget(permission, root, *parts))
+            val finalRoles = mutableMapOf<AccessTarget, UUID>()
+            if (targetsToNewIds.isEmpty()){
```

Comment:
not hugely important, but seems preferable to move this block up, just to avoid opening/closing a database connection in the case we don't need to

Reply:
good catch